File: MATLAB_models/geoTES_class.m
classdef geoTES_class
    properties
        mode % separate, continuous, or flowing
        recovery
        Tinit
        flowrate_per_well_prod % L/s
        flowrate_per_well_inj  % L/s
        Qmax
        depth
        thickness % m
        productivity
        injectivity

        nY

        rho
        cp
        void

        charge_production 
        charge_injection 
        discharge_production 
        discharge_injection 

        charge_prod_Nwell
        charge_inj_Nwell

        discharge_prod_Nwell
        discharge_inj_Nwell

        power
        energy
        net_energy

        energy_in_tot
        energy_out_tot
        eff

        energy_capacity % GWh-th
        mass
        volume
        side_length

    end

    methods
        function obj = geoTES_class(recovery,Tinit,flowrate,Qmax,props,fluid,mode,nY)
            obj.recovery = recovery ;
            obj.Tinit = Tinit ;
            obj.nY = nY ;
            if isstruct(flowrate)
                obj.flowrate_per_well_inj  = flowrate.I;
                obj.flowrate_per_well_prod = flowrate.P;
            else
                obj.flowrate_per_well_inj  = flowrate ;
                obj.flowrate_per_well_prod = flowrate ;
            end
            obj.Qmax = Qmax ;

            obj.rho = props.rho ;
            obj.cp = props.cp;
            obj.void = props.void;
            obj.depth = props.depth ;
            obj.thickness = props.thickness ;
            obj.productivity = props.productivity ;
            obj.injectivity = props.injectivity ;

            obj.mode = mode ;

            obj.charge_production = fluid_class(fluid) ;
            obj.charge_injection = fluid_class(fluid) ;
            obj.discharge_production = fluid_class(fluid) ;
            obj.discharge_injection = fluid_class(fluid) ;

            obj.power  = zeros(nY*8760,1) ;
            obj.energy = zeros(nY*8760,1) ;
        end

        % Calculate the physical size of the RTES
        function obj = geoTES_size(obj)
            if obj.charge_injection.T > obj.charge_production.T
                obj.energy_capacity = max(obj.energy((obj.nY-1)*8760+1:obj.nY*8760))/1000;
            else
                obj.energy_capacity = -min(obj.energy((obj.nY-1)*8760+1:obj.nY*8760))/1000;
            end
            en_cap_J = obj.energy_capacity * 3600e9 ;
            obj.mass = en_cap_J / (obj.cp * abs(obj.charge_injection.T - obj.charge_production.T)); % kg
            obj.volume = obj.mass / obj.rho / (1-obj.void) ;
            obj.side_length = obj.volume ^ (1/3);

            obj.net_energy = (obj.energy(obj.nY*8760) - obj.energy((obj.nY-1)*8760+1))/1000; % Net change in energy content of reservoir over one year, GWh-th
        end

        % Calculate flow rates and number of wells
        function obj = geoTES_well_flows(obj)
            if obj.charge_injection.T > obj.charge_production.T
                max_charge_power = max(obj.power((obj.nY-1)*8760+1:obj.nY*8760)) ;
                max_discharge_power = -min(obj.power((obj.nY-1)*8760+1:obj.nY*8760)) ;
            else
                max_charge_power = -min(obj.power((obj.nY-1)*8760+1:obj.nY*8760)) ;
                max_discharge_power = max(obj.power((obj.nY-1)*8760+1:obj.nY*8760)) ;
            end

            if max_charge_power == 0
                obj.charge_production.mdot = 0;
                obj.charge_injection.mdot = 0;
                obj.charge_production.vdot = 0;
                obj.charge_injection.vdot = 0;
            else
                obj.charge_production.mdot = max_charge_power * 1000 / abs(obj.charge_injection.h-obj.charge_production.h);
                obj.charge_injection.mdot = obj.charge_production.mdot ;
                obj.charge_injection = calc_mdot_vdot(obj.charge_injection,'vdot');
                obj.charge_production = calc_mdot_vdot(obj.charge_production,'vdot');
            end

            if max_discharge_power == 0
                obj.discharge_production.mdot = 0 ;
                obj.discharge_injection.mdot = 0;
                obj.discharge_production.vdot = 0 ;
                obj.discharge_injection.vdot = 0;
            else
                obj.discharge_production.mdot = max_discharge_power * 1000 / abs(obj.discharge_injection.h-obj.discharge_production.h);
                obj.discharge_injection.mdot = obj.discharge_production.mdot ;
                obj.discharge_injection = calc_mdot_vdot(obj.discharge_injection,'vdot');
                obj.discharge_production = calc_mdot_vdot(obj.discharge_production,'vdot');
            end

           
            % Calculate the number of wells
            obj.charge_prod_Nwell = ceil(obj.charge_production.vdot * 1000 / obj.flowrate_per_well_prod);
            obj.charge_inj_Nwell = ceil(obj.charge_injection.vdot * 1000 / obj.flowrate_per_well_inj);

            obj.discharge_prod_Nwell = ceil(obj.discharge_production.vdot * 1000 / obj.flowrate_per_well_prod);
            obj.discharge_inj_Nwell = ceil(obj.discharge_injection.vdot * 1000 / obj.flowrate_per_well_inj);


        end

        function obj = geoTES_annual_energy(obj)

            n = (obj.nY-1)*8760+1:obj.nY*8760 ; 
            pow = obj.power(n) ;
            obj.energy_in_tot = sum(pow(pow>0)) / 1000 ;
            obj.energy_out_tot = -sum(pow(pow<0)) / 1000;
            obj.eff = obj.energy_out_tot / obj.energy_in_tot ;
        end
    end

end

File: Python_TEA_model/geoTES_class.py
import math

import numpy as np

from fluid_class import fluid_class


class geoTES_class:
    def __init__(self, recovery, Tinit, flowrate, Qmax, props, fluid, nY=2):
        self.recovery = float(recovery)
        self.Tinit = float(Tinit)
        self.flowrate_per_well = float(flowrate)
        self.Qmax = float(Qmax)
        self.nY = int(nY)

        self.rho = float(props["rho"])
        self.cp = float(props["cp"])
        self.void = float(props["void"])

        self.charge_production = fluid_class(fluid)
        self.charge_injection = fluid_class(fluid)
        self.discharge_production = fluid_class(fluid)
        self.discharge_injection = fluid_class(fluid)

        self.charge_prod_Nwell = 0
        self.charge_inj_Nwell = 0
        self.discharge_prod_Nwell = 0
        self.discharge_inj_Nwell = 0

        self.power = np.zeros(self.nY * 8760, dtype=float)
        self.energy = np.zeros(self.nY * 8760, dtype=float)

        self.energy_in_tot = 0.0
        self.energy_out_tot = 0.0
        self.eff = 0.0
        self.energy_capacity = 0.0
        self.mass = 0.0
        self.volume = 0.0
        self.side_length = 0.0
        self.net_energy = 0.0

    def geoTES_size(self):
        hours = 8760
        hours2 = self.nY * hours
        start = (self.nY - 1) * hours

        if self.charge_injection.T > self.charge_production.T:
            self.energy_capacity = np.max(self.energy[start:hours2]) / 1000.0
        else:
            self.energy_capacity = -np.min(self.energy[start:hours2]) / 1000.0

        en_cap_J = self.energy_capacity * 3600e9
        self.mass = en_cap_J / (self.cp * abs(self.charge_injection.T - self.charge_production.T))
        self.volume = self.mass / self.rho / (1.0 - self.void)
        self.side_length = self.volume ** (1.0 / 3.0)
        self.net_energy = (self.energy[hours2 - 1] - self.energy[(self.nY - 1) * hours]) / 1000.0

        return self

    def geoTES_well_flows(self):
        hours = 8760
        hours2 = 2 * hours

        if self.charge_injection.T > self.charge_production.T:
            max_charge_power = np.max(self.power[hours:hours2])
            max_discharge_power = -np.min(self.power[hours:hours2])
        else:
            max_charge_power = -np.min(self.power[hours:hours2])
            max_discharge_power = np.max(self.power[hours:hours2])

        if max_charge_power == 0.0:
            self.charge_production.mdot = 0.0
            self.charge_injection.mdot = 0.0
        else:
            self.charge_production.mdot = (
                max_charge_power * 1000.0 / abs(self.charge_injection.h - self.charge_production.h)
            )
            self.charge_injection.mdot = self.charge_production.mdot

        self.charge_injection.calc_mdot_vdot("vdot")
        self.charge_production.calc_mdot_vdot("vdot")

        if max_discharge_power == 0.0:
            self.discharge_production.mdot = 0.0
            self.discharge_injection.mdot = 0.0
        else:
            self.discharge_production.mdot = (
                max_discharge_power * 1000.0 / abs(self.discharge_injection.h - self.discharge_production.h)
            )
            self.discharge_injection.mdot = self.discharge_production.mdot

        self.discharge_injection.calc_mdot_vdot("vdot")
        self.discharge_production.calc_mdot_vdot("vdot")

        self.charge_prod_Nwell = int(
            math.ceil(self.charge_production.vdot * 1000.0 / self.flowrate_per_well)
        )
        self.charge_inj_Nwell = int(
            math.ceil(self.charge_injection.vdot * 1000.0 / self.flowrate_per_well)
        )
        self.discharge_prod_Nwell = int(
            math.ceil(self.discharge_production.vdot * 1000.0 / self.flowrate_per_well)
        )
        self.discharge_inj_Nwell = int(
            math.ceil(self.discharge_injection.vdot * 1000.0 / self.flowrate_per_well)
        )

        return self

    def geoTES_annual_energy(self):
        hours = 8760
        hours2 = self.nY * hours
        start = (self.nY - 1) * hours
        pow_slice = self.power[start:hours2]

        self.energy_in_tot = float(np.sum(pow_slice[pow_slice > 0]) / 1000.0)
        self.energy_out_tot = float(-np.sum(pow_slice[pow_slice < 0]) / 1000.0)
        self.eff = self.energy_out_tot / self.energy_in_tot if self.energy_in_tot != 0 else 0.0

        return self
